Aydin nilay/issue12

niloaydin opened the pull request:
**Local Authentication for Login & Crud operation on /profile**

- Created a local authentication for `log in route` with JWT in auth controller. 
- Created 7 tests for `/login` and `/profile` routes using supertest under `__tests__ folder` in `auth.spec.test file`
- Added an `error handling middleware` in `app.js`
- Created validation middleware for `login` which checks if there is an email and password, and if the email format is correct
- Created 2 swagger documentation: One for `/login` endpoint and one for `profile` endpoint under `doc folder` in `profile.yaml` and `login.yaml`

niloaydin committed:
Initial set up

niloaydin committed:
1.Created /login and /profile endpoint in auth controller
2.Add errorhandler in app.js
3.Create validation middleware for login
4.Create tests for login and profile endpoints
5.Create swagger documents for login and profile

niloaydin committed:
Change the test script in package.json

khatibAmjad commented on src/routes/auth.js:
```diff
@@ -1,6 +1,18 @@
+  errorHandlingForValidation,
+  authController.loginUsers
+);
+router.get('/profile', authController.getProfile);
```
I think it makes more sense for the profile router to be in the `/api/use/` endpoint.

niloaydin replied:
I was actually confused about this, but in some of the websites, I saw that it was under auth. But lets open this to discussion

khatibAmjad commented on src/app.js:
```diff
@@ -14,15 +19,36 @@ const postRoutes = require('./routes/post');
+  '/api/about',
+  '/api/auth/login',
+  '/api/auth/register',
+];
```
since this middleware is gonna be used on all the endpoints can you add the other public endpoints to the array, like `/`, `/api/auth/google`, `/api/auth/google/callback`, `/api/auth/facebook`, `/api/auth/facebook/callback` and put it in the constant file.
![image](https://user-images.githubusercontent.com/98318641/182239449-d9e775b2-661d-4728-bdfb-563114f44b0f.png)

niloaydin replied:
sure

khatibAmjad commented on src/controllers/auth.js:
```diff
@@ -1,2 +1,53 @@
+};
+
+// eslint-disable-next-line consistent-return
+const getProfile = async (req, res) => {
```
if the router will be in the user routes this'll be in the user controllers too ofc

khatibAmjad commented on src/__tests__/auth.spec.js:
```diff
@@ -1,5 +1,152 @@
+        .set('Content-Type', 'application/json')
+        .send({
+          email: validUserExample.email,
+          password: 'wrongpasswordthaticreatedfortestpurposes',
```
Why did you create this wrong password?

niloaydin replied:
Because I am testing if the code is gonna return an error if there is a wrong password passed

khatibAmjad commented on src/app.js:
```diff
@@ -57,8 +83,28 @@ app.get('/', (req, res) => {
   res.send("Hello Let's Share!");
 });
 
+// If I delete the next, the code is not working. Why?
```
idk...

niloaydin replied:
I cry

khatibAmjad reviewed (commented):
The work is super impressive kankam

niloaydin committed:
1.Delete yarn.lock and install dependencies again
2.Move /profile endpoint from auth to user

niloaydin committed:
Change the profile route from auth to user in test

niloaydin committed:
Delete the unused code

halitbatur commented on src/controllers/auth.js:
```diff
@@ -1,2 +1,37 @@
+    const databaseUser = await UserModel.findOne({ email });
+
+    if (!databaseUser) {
+      return res.status(401).json({ message: 'Wrong username or password!' });
```
can we fix the error message here?

halitbatur commented on src/controllers/auth.js:
```diff
@@ -1,2 +1,37 @@
+    // eslint-disable-next-line no-underscore-dangle
+    const payload = { userId: databaseUser._id };
+    const token = jwt.sign(payload, process.env.SECRET_KEY, {
+      expiresIn: '14d',
```
can we not use magic strings please, lets put this in a variable in our const file and import it from there

halitbatur commented on src/controllers/auth.js:
```diff
@@ -1,2 +1,37 @@
+    });
+    res.cookie('token', token, {
+      httpOnly: true,
+      maxAge: 1000 * 60 * 60 * 24 * 14,
```
Can we move this to a variable, and write a comment of how long it actually takes for the token to expire

halitbatur commented on src/controllers/auth.js:
```diff
@@ -1,2 +1,37 @@
+const jwt = require('jsonwebtoken');
 const UserModel = require('../models/user');
+
+const loginUsers = async (req, res) => {
```
can we change this to `login`

halitbatur commented on src/controllers/auth.js:
```diff
@@ -1,2 +1,37 @@
+const loginUsers = async (req, res) => {
+  try {
+    const { email, password } = req.body;
+    const databaseUser = await UserModel.findOne({ email });
```
can we change this to `currentUser`

halitbatur commented on src/docs/login.yaml:
```diff
@@ -0,0 +1,25 @@
+    parameters:
+    - name: "email"
+      in: "body"
+      description: "Esdfsdfsdlogin"
```
whats that?

halitbatur commented on src/docs/login.yaml:
```diff
@@ -0,0 +1,25 @@
+      description: "Password for login"
+      required: true
+      type: "string"
+    responses:
```
"403" status code is not here

halitbatur commented on src/docs/profile.yaml:
```diff
@@ -0,0 +1,13 @@
+        "200":
+          description: "User info is fetched"
+        "401":
+          description: "UnauthorizedError: No authorization token was found"
```
can u add a new line at the end of the file

halitbatur commented on src/docs/profile.yaml:
```diff
@@ -0,0 +1,13 @@
+      responses:
+        "200":
+          description: "User info is fetched"
+        "401":
```
403*

halitbatur commented on src/middleware/validation.js:
```diff
@@ -0,0 +1,21 @@
+  const errors = validationResult(req);
+
+  if (!errors.isEmpty()) {
+    return res.status(422).json({ errors: errors.array() });
```
can we add this to swagger too

halitbatur commented on src/routes/post.js:
```diff
@@ -1,5 +1,6 @@
 const express = require('express');
-const postController = require('../controllers/post');
+
+// const postController = require('../controllers/post');
```
no commented code please

halitbatur reviewed (changes requested):
Good work! some few changes and its ready to be merged

halitbatur commented:
Good description @niloaydin Keep up the good work!

niloaydin committed:
1.Add missing status codes on login and profile yaml files
2.Change the name of the function to login and the databaseUser to currentUser
3.Change the error status from 403 to 500 and delete the error message in user controller
4.Add new constanst in contanst file and use them in the controller
5.Delete the commentouted lines

halitbatur commented on src/controllers/user.js:
```diff
@@ -0,0 +1,23 @@
+    }
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.log(error);
```
can we remove this? and pass the error with the response

halitbatur commented on src/lib/constants.js:
```diff
@@ -33,4 +33,7 @@ module.exports = Object.freeze({
   ],
   ENUM_POST_TYPE: ['Request', 'Donate'],
   ENUM_POST_STATUS: ['Published', 'Verified'],
+
```
extra new line

halitbatur reviewed (approved):
Looks good Nilay thank you for the quick changes, make sure to do the changes i commented on and its ready to be merged

halitbatur commented:
@niloaydin oh and conflicts needs to be resolved

niloaydin committed:
Add error message in user controller

niloaydin committed:
Merge Amjad's branch and fix the merge conflicts

niloaydin committed:
Delete UserModel collection after every test

niloaydin committed:
Add google auth paths to publick paths

niloaydin committed:
Fix the issues in tests

niloaydin committed:
Change some minor issues

niloaydin committed:
1.fix the issues in constans in lib
2.Applied those changed in auth controller
3.Merge auth.spec with amjad's tests
